### DISCUSSION TITLE
How to include plugin libs in classpath for Java API properly ?

### DISCUSSION THREAD

**Discussion by TiboGr** [dita-ot/dita-ot#4373] [Q&A]
Hello, 
I'm using the java API to run Dita-OT from a java application.
When running pdf transtype, i'm having issues with some ClassNotFoundExceptions.
I have dita-ot/config and dita-ot/lib directories in my classpath. 

Some plugins like **org.dita.index** and **org.dita.pdf2.fop** do not reference their jars.

Some like **org.dita.pdf2** do this and it works fine 
```
    <path id="project.class.path">
      <pathelement path="${java.class.path}"/>
      <pathelement location="${fo.lib.dir}/fo.jar"/>
    </path>
...
<taskdef name="detect-lang" classname="com.suite.sol.ditaot.DetectLang" classpathref="project.class.path"/>
```

I also found that it used to be the case for fop at least (https://github.com/dita-ot/dita-ot/commit/ffcb7b19c6c6ee1ae943c50b33524eafd15bb4e7)
Some issues mention that i should include all plugins jar libs in my classpath (https://github.com/dita-ot/dita-ot/issues/3647#issuecomment-749175293), but is it the best way ? Shouldn't plugins always reference their libs ?

**Comment by raducoravu** [comment]
If you look in the file "DITA-OT/config/env.bat", the DITA OT gathers all JAR library references from all plugins and creates a batch script which is run when running the "dita" executable. So the DITA OT itself does not have its own class loader which would load all libraries from certain folders and then start the processing.
I would say that you should include all your plugin JARs in the Java class path.
Better yet, as someone working for a company with a Java application which publishes using the DITA OT, we start from our Java code the "dita" executable directly as a separate process, this gives us several advantages:
 - The DITA OT remains like a black box which loads all the appropriate libraries, we just send parameters to the "dita" script and intercept the console out/err streams of the started process.
 - If the DITA OT has an internal failure or an out of memory, our application does not suffer from that.
